backend/src/LibExecution/Interpreter.fs

```diff
@@ -633,14 +633,23 @@ let rec eval
 
       let mutable matchResult = None
 
-      for (pattern, rhsExpr) in NEList.toList cases do
+      for case in NEList.toList cases do
         if Option.isSome matchResult then
           ()
         else
-          let! passes, newDefs = checkPattern matchVal pattern
+          let! passes, newDefs = checkPattern matchVal case.pat
           let newSymtable = Map.mergeFavoringRight st (Map.ofList newDefs)
-          if matchResult = None && passes then
-            let! r = eval state tlid tst newSymtable rhsExpr
+          let! whenCondition =
+            uply {
+              match case.whenCondition with
+              | Some whenCondition ->
+                match! eval state tlid tst newSymtable whenCondition with
+                | DBool b -> return b
+                | _ -> return false
+              | None -> return true
+            }
```

Comment:
This worked out tidily! Maybe rename from `whenCondition` to `passesWhenCondition`?
and rename `passes` to `passesPattern`?